### DISCUSSION TITLE
Can i build a custom check for cloud ressources in renovate?

### DISCUSSION THREAD

**Discussion by bok11** [renovatebot/renovate#22734] [Request Help]
### How are you running Renovate?

Placeholder value, please select the correct response from the dropdown

### If you're self-hosting Renovate, tell us what version of Renovate you run.

newest NPX

### If you're self-hosting Renovate, select which platform you are using.

Azure DevOps (dev.azure.com)

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

I have a repository deploying infrastructure as code (biceps template)
It has a variable called "kubernetesversion", and I need to sync this to the latest version available in azure.

Is it possible to do something custom like this with renovate, or should i write a seperate script?

### What you tried so far.

Google, and looking in documentation

### Relevant debug logs

<details><summary>Logs</summary>

```
Copy/paste the relevant log(s) here, between the starting and ending backticks
```

</details>

**Comment by rarkins** [comment]
> the latest version available in azure

Where/how is that version available? e.g. is there an API?

**Reply by bok11** [reply]
Yes, it can be queried from their rest API (requires authentication): https://learn.microsoft.com/en-us/rest/api/aks/managed-clusters/list-kubernetes-versions?tabs=HTTP#kubernetesversionlistresult

**Reply by bok11** [reply]
Also, we can get it with the azure cli, here is an example output.

In the end I want to find the current latest version (1.26.3)  and make sure that is what I have in my bicep file as "kubernetesversion = 1.26.3"

This could also be relevant for database versions or similar.

```
>> az aks get-versions --location westeurope
{
  "id": "/subscriptions/<subscriptionid>/providers/Microsoft.ContainerService/locations/westeurope/orchestrators",
  "name": "default",
  "orchestrators": [
    {
      "default": null,
      "isPreview": null,
      "orchestratorType": "Kubernetes",
      "orchestratorVersion": "1.24.9",
      "upgrades": [
        {
          "isPreview": null,
          "orchestratorType": "Kubernetes",
          "orchestratorVersion": "1.24.10"
        },
        {
          "isPreview": null,
          "orchestratorType": "Kubernetes",
          "orchestratorVersion": "1.25.5"
        },
        {
          "isPreview": null,
          "orchestratorType": "Kubernetes",
          "orchestratorVersion": "1.25.6"
        }
      ]
    },
    {
      "default": null,
      "isPreview": null,
      "orchestratorType": "Kubernetes",
      "orchestratorVersion": "1.24.10",
      "upgrades": [
        {
          "isPreview": null,
          "orchestratorType": "Kubernetes",
          "orchestratorVersion": "1.25.5"
        },
        {
          "isPreview": null,
          "orchestratorType": "Kubernetes",
          "orchestratorVersion": "1.25.6"
        }
      ]
    },
    {
      "default": null,
      "isPreview": null,
      "orchestratorType": "Kubernetes",
      "orchestratorVersion": "1.25.5",
      "upgrades": [
        {
          "isPreview": null,
          "orchestratorType": "Kubernetes",
          "orchestratorVersion": "1.25.6"
        },
        {
          "isPreview": null,
          "orchestratorType": "Kubernetes",
          "orchestratorVersion": "1.26.0"
        },
        {
          "isPreview": null,
          "orchestratorType": "Kubernetes",
          "orchestratorVersion": "1.26.3"
        }
      ]
    },
    {
      "default": true,
      "isPreview": null,
      "orchestratorType": "Kubernetes",
      "orchestratorVersion": "1.25.6",
      "upgrades": [
        {
          "isPreview": null,
          "orchestratorType": "Kubernetes",
          "orchestratorVersion": "1.26.0"
        },
        {
          "isPreview": null,
          "orchestratorType": "Kubernetes",
          "orchestratorVersion": "1.26.3"
        }
      ]
    },
    {
      "default": null,
      "isPreview": null,
      "orchestratorType": "Kubernetes",
      "orchestratorVersion": "1.26.0",
      "upgrades": [
        {
          "isPreview": null,
          "orchestratorType": "Kubernetes",
          "orchestratorVersion": "1.26.3"
        }
      ]
    },
    {
      "default": null,
      "isPreview": null,
      "orchestratorType": "Kubernetes",
      "orchestratorVersion": "1.26.3",
      "upgrades": null
    }
  ],
  "type": "Microsoft.ContainerService/locations/orchestrators"
}
```

**Reply by secustor** [reply]
No there is no direct way. 

You can write your own custom datasource, tough there is a generic solution in the works. https://github.com/renovatebot/renovate/issues/10622

**Reply by bok11** [reply]
Thanks for the response! 
Are there any documentation on adding a custom data source?
And alternatively, are you accepting pull requests if I want to proceed on your the work in your branch for the generic datasource?

**Reply by secustor** [reply]
> Are there any documentation on adding a custom data source?

No, because we do not give stability guarantees for internal APIs we consider this outside of the normal usage of Renovate.    
You basically have to fork Renovate, implement the datasource and then build your custom image. 
If you plan to contribute the datasource back, we should discuss it first here and create an issue. 

> And alternatively, are you accepting pull requests if I want to proceed on your the work in your branch for the generic datasource?

Yeah, sadly I had not much time to work on it and lacking a sponsor.    

I would accept additions, tough the PR needs a complete rewrite as the requirements have changed. https://github.com/renovatebot/renovate/pull/20753#issuecomment-1562398864.   
Also it now requires changes in multiple systems of Renovate which makes a little bit much for a first time contributor, but I'm surely not preventing you from trying. ;)
